File: rails/app/services/marketplace/tradelisting_service.rb
module Marketplace
  class TradelistingService
    def initialize(repository = Tradelisting)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def close(id)
      instance = Tradelisting.find(id)
      instance.close()
      instance.save!
    end

    def extend(id, days)
      instance = Tradelisting.find(id)
      instance.extend(days)
      instance.save!
    end

    def cancel(id)
      instance = Tradelisting.find(id)
      instance.cancel()
      instance.save!
    end

    # triggered by @on(status = Sold)
    def set_status(id, value)
      instance = Tradelisting.find(id)
      instance.status = value
      if value.to_s.upcase == 'SOLD'
        instance.finalize_auction
      end
      instance.save!
    end
  end
end

File: laravel/app/Services/Marketplace/TradelistingService.php
<?php

namespace App\Services\Marketplace;

use App\Models\Marketplace\TradeListing;

class TradeListingService
{
    public function create(array $data): TradeListing
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TradeListing $tradeListing, array $data): TradeListing
    {
        throw new \LogicException('Not implemented');
    }
    public function close(int $id): void
    {
        $tradeListing = TradeListing::findOrFail($id);
        $tradeListing->close();
        $tradeListing->save();
    }

    public function extend(int $id, $days): void
    {
        $tradeListing = TradeListing::findOrFail($id);
        $tradeListing->extend($days);
        $tradeListing->save();
    }

    public function cancel(int $id): void
    {
        $tradeListing = TradeListing::findOrFail($id);
        $tradeListing->cancel();
        $tradeListing->save();
    }

    public function isExpired(int $id): bool
    {
        $tradeListing = TradeListing::findOrFail($id);
        $result = $tradeListing->isExpired();
        $tradeListing->save();
        return $result;
    }

    public function finalizeAuction(int $id): void
    {
        $tradeListing = TradeListing::findOrFail($id);
        $tradeListing->finalizeAuction();
        $tradeListing->save();
    }

    // triggered by @on(status = Sold)
    public function setStatus(int $id, string $value): void
    {
        $tradeListing = TradeListing::findOrFail($id);
        $tradeListing->status = $value;
        if ($value === 'Sold') {
            $tradeListing->finalizeAuction();
        }
        $tradeListing->save();
    }
}
